TE-18.1 gRIBI MPLS in UDP encap and decap

dplore opened the pull request:
Test specification for gRIBI encap/decap with an input-policy which classifies and polices incoming packets using a one-rate, two color marker.

Several TODO items exist to acknowledge planned additions to the test.

## Highlights
* introduced new OC subtree (/qos/input-policies) 
  * The operational use case is to configure schedulers (policers) on input subinterfaces
  * Today OC requires definition of queues for schedulers but in practice, input policing typically does not involve queues
  * A new `input-policies` subtree is proposed here, relating classifiers and schedulers to each other.
  * `input-policies` are then applied to an interface with a new OC leaf-list (qos/interfaces/interface/input/policies)

Related PR's:
* #3375 defines scaling requirements for this functionality.
* #3383 defines a policer related to this test
*

dplore commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,433 @@
+* Each policer is assigned rate limits matching one of 800 different possible limits between 1Gbps to 400Gbps in 0.5Gbps increments
+
+Note that inner IP address space is reused / overlapping between network-instances.
+
```
add frequency of gRIBI and gNMI updates

dplore replied:
Moved to https://github.com/openconfig/featureprofiles/pull/3375

vishnureddybadveli reviewed (commented):
Would be great to see if multi vendor can support TE-18.1.1, TE-18.1.2, TE-18.1.3, TE-18.1.4

vishnureddybadveli reviewed (approved):
Overall changes are good.

dplore commented:
> Overall changes are good.


@vishnureddybadveli see latest change at: https://github.com/openconfig/featureprofiles/pull/3363/files/bfcadd11237ec96f7872a88a8062346d74f55695..b5f414856305b5b065d61067598b3d80baa678e8

- I added a commit correcting the UDP to GRE encap for the default encap gRIBI entry.  Also I couldn't remember if we wanted gnmi config or gRIBI for the default encap (I know we want gnmi config for the default decap and that's in the README now).  

- refactored the minimum ttl rule into a gRIBI parameter for the mpls-in-udp encap.  This seemed more than using a qos configuration or adding in some policy-forwarding config.

LimeHat commented:
> introduced new OC subtree (/qos/input-policies)
The operational use case is to configure schedulers (policers) on input subinterfaces
Today OC requires definition of queues for schedulers but in practice, input policing typically does not involve queues
A new input-policies subtree is proposed here, relating classifiers and schedulers to each other.
input-policies are then applied to an interface with a new OC leaf-list (qos/interfaces/interface/input/policies)

Is there an open PR for this already? (I wasn't able to find it)

I would argue that this may not be an optimal solution. Overloading the scheduler-policies, which are fundamentally responsible for queue management with additional, unrelated functionality (definition/instantiation of ingress policers on subinterfaces) is questionable.

dplore commented:
> > introduced new OC subtree (/qos/input-policies)
> > The operational use case is to configure schedulers (policers) on input subinterfaces
> > Today OC requires definition of queues for schedulers but in practice, input policing typically does not involve queues
> > A new input-policies subtree is proposed here, relating classifiers and schedulers to each other.
> > input-policies are then applied to an interface with a new OC leaf-list (qos/interfaces/interface/input/policies)
> 
> Is there an open PR for this already? (I wasn't able to find it)
> 
> I would argue that this may not be an optimal solution. Overloading the scheduler-policies, which are fundamentally responsible for queue management with additional, unrelated functionality (definition/instantiation of ingress policers on subinterfaces) is questionable.

Thanks for the feedback, there is no open PR for the proposed OC model in this README.  Note the policer configuration has been moved to https://github.com/openconfig/featureprofiles/pull/3383.  Let's discuss there.

trlongth commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,395 @@
+Create AFT entries using gRIBI to match destination IP address in a
+network-instance.  Encapsulate the matching packets in MPLS in UDP. Configure
+a qos scheduler using gNMI to rate limit / police traffic based on matching
+the destination IP address of a packet input to the DUT.
```
Need to update to matching based on next-hop group instead of destination IP address of a packet input.

dplore replied:
done

vishnureddybadveli commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,395 @@
+network-instance.  Encapsulate the matching packets in MPLS in UDP. Configure
+a qos scheduler using gNMI to rate limit / police traffic based on matching
+the destination IP address of a packet input to the DUT.
+
```
Can encap all packets received for a VRF and send to a statically configured destination node (NHG), policy forwarding?

dplore replied:
revised

trlongth commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,395 @@
+network-instance.  Encapsulate the matching packets in MPLS in UDP. Configure
+a qos scheduler using gNMI to rate limit / police traffic based on matching
+the destination IP address of a packet input to the DUT.
+
```
Add encapsulation for ethernet packets with mpls-in-udp IPv6 (in a separate README).

Encap all packets received in a VRF and send to statically configured IP (up to 16) (NHG) - ie: this is policy-forwarding in OC.

vishnureddybadveli replied:
> Add encapsulation for ethernet packets with mpls-in-udp IPv6 (in a separate README).
> 
> Encap all packets received in a VRF and send to statically configured IP (up to 16) (NHG) - ie: this is policy-forwarding in OC.

I think above comment refers the same.

dplore replied:
done

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,378 @@
+# TE-18.1 gRIBI MPLS in UDP Encapsulation with OC QoS scheduler
+
+Create AFT entries using gRIBI to match on next hop group in a
+network-instance and encapsulate the matching packets in MPLS in UDP.
```
MPLS in UDP IPv6

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,378 @@
+
+TODO: Complete test environment setup steps
+
+inner_ipv6_dst_A = "2001:aa:bb::1/128"
```
q: tbd w/ vendors: how do the BGP routing work w/o the PBR solution ? what if the inner dst_A belongs to CR and used to traverse BGP packets ? Currently the PBR rules intercept only echo reply packets and all other packets are encapsulated. w/ vrf there will not be any PBR rule. Not sure if this is discussed and sorted in past

dplore replied:
Added TODO item to specify a solution for ensuring BGP packets are matched, encapsulated and forwarding to a specified  destination using OC policy-forwarding terms.

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,378 @@
+    afts {
+      ipv6_unicast {
+        ipv6_entry {
+          prefix: "outer_loopback_ipv6"   # IPv6 match rule for the device loopback expected to receive MPLS in UDP packets
```
it can be loopback or any range configured on the router.

dplore replied:
by range, I assume you mean prefix?  Fixed.

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,378 @@
+          decapsulate_header: OPENCONFIG_AFT_TYPES:MPLS_IN_UDPV6      # TODO: Add OC path for this
+          # The device should decapsulate the UDP packet and use the MPLS header
+          # to determine the destination network-instance, then forward the packet
+          # based on the inner IP packet, matching an appropriate AFT entry.
```
This needs to be updated: 

# The device should decapsulate the UDP packet and use the MPLS header
# to determine the outgoing interface and then forward the packet on external interface
# without checking the inner IP packet.

dplore replied:
done

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,378 @@
+
+### TE-18.1.5 Rewrite inner packet TTL=2 if inner TTL=1
+
+* Perform mpls-in-udp encapsulation, but add a condition that the DUT must
```
or set to 1. In short if the incoming packet TTL is 1, the same value 1 should go out. setting 2 will set the ttl to 2.

vishnureddybadveli replied:
Darren Loher,
Could you please remove TTL modifications from the pull request.
We can continue discussion at #3409
https://github.com/openconfig/featureprofiles/pull/3409

dplore replied:
Done

vishnureddybadveli commented:
decap rules to use policy-forwarding changes looks good.

vishnureddybadveli commented:
Changes looks good to me.

dplore commented:
@trlongth  @jaingaurav12 all comments addressed.  Please review, thanks.

vishnureddybadveli reviewed (approved):
Changes are good to me.

dplore commented:
I made a couple of corrections and I think this is looking good now with no planned changes.

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,324 @@
+
+Create AFT entries using gRIBI to match on next hop group in a
+network-instance and encapsulate the matching packets in MPLS in UDP.
+
```
Create a default AFT entry using gNMI to match on next hop group in a
network-instance and encapsulate the matching packets in MPLS in GRE IPv4.

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,324 @@
+# TE-18.1 gRIBI MPLS in UDP Encapsulation and Decapsulation
+
+Create AFT entries using gRIBI to match on next hop group in a
+network-instance and encapsulate the matching packets in MPLS in UDP.
```
MPLS in UDP. ---> MPLS in UDP IPv6

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,324 @@
+network-instance and encapsulate the matching packets in MPLS in UDP.
+
+Create a policy routing configuration using gNMI to decapsulate MPLS
+in UDP packets which are sent to a loopback address and apply to
```
Pls change this 

From: "to decapsulate MPLS in UDP packets " 
to: "to decapsulate MPLS in UDP and MPLS in GRE packets " 

(anPF needs to decapsulate both type of packets. One coming from Zakim and other from the VM host

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,324 @@
+                rule: 1
+                  config:
+                    sequence-id: 1
+                  ipv6:
```
IPv6 or Ipv4.
Currently only v4 is supported by the backend system but in coming future, we may shift to v6

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,324 @@
+                            destination-ip: "outer_ipv6_dst_def"
+                            ip-ttl: outer_ip-ttl
+                            dscp: outer_dscp
+                            inner-ttl-min: 2
```
Darren, This is a condition only when incoming packet (coming from customer) TTL is 1 and we want the (inner) packet to go out w/ ttl 1. shall we clarify the PBR policy to have two stage implementation. thank you

jaingaurav12 commented on feature/gribi/otg_tests/mpls_in_udp/README.md:
```diff
@@ -0,0 +1,324 @@
+                    sequence-id: 1
+                  ipv6:
+                    config:
+                      destination-address: "decap_loopback_ipv6"
```
"destination-address: "decap_loopback_ipv6"
This is a range of IPs. Same is the scenario during the encap where MPLS o GRE sources are from the decap range
